src/Common/src/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -1797,7 +1797,17 @@ protected override void TransformNativeToManaged(ILCodeStream codeStream)
                 Context.GetHelperEntryPoint("InteropHelpers", "GetPInvokeDelegateForStub")));
 
             StoreManagedValue(codeStream);
+        }
 
+        protected override void EmitCleanupManaged(ILCodeStream codeStream)
+        {
+            if (In
+                && MarshalDirection == MarshalDirection.Forward
+                && MarshallerType == MarshallerType.Argument)
+            {
+                LoadManagedValue(codeStream);
+                codeStream.Emit(ILOpcode.call, _ilCodeStreams.Emitter.NewToken(InteropTypes.GetGC(Context).GetKnownMethod("KeepAlive", null)));
```

Comment:
Thinking about this a bit more, we should only need to emit the GC.KeepAlive call in the unmarshal stream. CoreCLR does the same thing.